examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2893 @@
+    // concerned about reading and writing matrix and vector entries. They
+    // are mainly motivated by providing slightly more efficient code and
+    // <a href="https://en.wikipedia.org/wiki/Syntactic_sugar"> syntactic
+    // sugar</a> for otherwise somewhat tedious code.
+
+    // The first function we introduce, <code>get_entry()</code>, will be
+    // used to read the value stored at the entry pointed by a
+    // SparsityPattern iterator <code>it</code> of <code>matrix</code>. The
+    // function works around a small deficiency in the SparseMatrix
+    // interface: The SparsityPattern is concerned with all index
+    // operations of the sparse matrix stored in CRS format. As such the
+    // iterator already knows the global index of the corresponding matrix
+    // entry in the low-level vector stored in the SparseMatrix object. Due
+    // to the lack of an interface in the SparseMatrix for accessing the
+    // element directly with a SparsityPattern iterator, we unfortunately
+    // have to create a temporary SparseMatrix iterator. We simply hide
+    // this in the <code>get_entry()</code> function.
+
+    template <typename Matrix, typename Iterator>
+    DEAL_II_ALWAYS_INLINE inline typename Matrix::value_type
+    get_entry(const Matrix &matrix, const Iterator &it)
+    {
+      const auto                            global_index = it->global_index();
+      const typename Matrix::const_iterator matrix_iterator(&matrix,
+                                                            global_index);
+      return matrix_iterator->value();
+    }
+
+    // The <code>set_entry()</code> helper is the inverse operation of
+    // <code>get_value()</code>: Given an iterator and a value, it sets the
+    // entry pointed to by the iterator in the matrix.
+
+    template <typename Matrix, typename Iterator>
+    DEAL_II_ALWAYS_INLINE inline void
+    set_entry(Matrix &                    matrix,
+              const Iterator &            it,
+              typename Matrix::value_type value)
+    {
+      const auto                global_index = it->global_index();
+      typename Matrix::iterator matrix_iterator(&matrix, global_index);
+      matrix_iterator->value() = value;
+    }
+
+    // <code>gather_get_entry()</code>: we note that $\mathbf{c}_{ij} \in
+    // \mathbb{R}^d$. If $d=2$ then $\mathbf{c}_{ij} =
+    // [\mathbf{c}_{ij}^1,\mathbf{c}_{ij}^2]^\top$. Which basically implies
+    // that we need one matrix per space dimension to store the
+    // $\mathbf{c}_{ij}$ vectors. Similar observation follows for the
+    // matrix $\mathbf{n}_{ij}$. The purpose of
+    // <code>gather_get_entry()</code> is to retrieve those entries a store
+    // them into a <code>Tensor<1, dim></code> for our convenience.
+
+    template <typename T1, std::size_t k, typename T2>
+    DEAL_II_ALWAYS_INLINE inline Tensor<1, k>
+    gather_get_entry(const std::array<T1, k> &U, const T2 it)
```

Comment:
Are you using this for general `T1`, or do you specifically have `SparseMatrix` in mind? If the latter, maybe just avoid the difficulty of reading the general template and not make the function generic.

Separately, let's use more descriptive template names -- say `SparseMatrixType` and `IteratorType`.